crypto/signals.c

```diff
@@ -0,0 +1,228 @@
+{
+    if (list_empty(&siglist))
+        return FALSE;
+
+    struct list *l;
+    if ((l = list_find(&siglist, siglist_sigeq, (void*)&dwType)) == NULL)
+        return FALSE;
+
+    container_of(l, struct sigentry, list)->p.callback(dwType);
+    return TRUE;
+}
+
+static int CRYPTO_SIGNAL_arch_enable()
+{
+    callback_handler = (PHANDLER_ROUTINE) CRYPTO_SIGNAL_handler;
+    if (SetConsoleCtrlHandler(callback_handler, TRUE) == 0)
+        return 0;
+    return 1;
+}
+
+static int CRYPTO_SIGNAL_arch_disable()
+{
+    if (SetConsoleCtrlHandler(callback_handler, FALSE) == 0) {
+        callback_handler = NULL;
+        return 0;
+    }
+    return 1;
+}
+
+static int CRYPTO_SIGNAL_block_arch(CRYPTO_SIGNAL *p)
+{
+    return 1;
+}
+
+#elif defined(OPENSSL_SYS_UNIX) || defined (__unix__) || \
+      (defined (__APPLE__) && defined (__MACH__))
+
+# include <sys/types.h>
+# include <unistd.h>
+# include <signal.h>
+
+static int CRYPTO_SIGNAL_arch_enable(void)
+{
+    return 1;
+}
+
+static int CRYPTO_SIGNAL_arch_disable(void)
+{
+    return 1;
+}
+
+/**
+ * signal specifies which signal to (un)block/(un)mask
+ * callback is one of:
+ *     NULL, SIG_DFL   restores default signal handling
+ *     SIG_IGN         blocks signal (if possible)
+ *     void (*)(int)   masks signal and calls a callback upon receive
+ */
+static int CRYPTO_SIGNAL_block_arch(CRYPTO_SIGNAL *p)
+{
+    int how;
+    sigset_t sigs;
+    struct sigaction sa;
+
+    memset(&sa, 0, sizeof(struct sigaction));
+    sa.sa_handler = p->callback;
+
+    if (sigemptyset (&sa.sa_mask) != 0)
+        goto fail;
+
+    if (sigaction((int)p->signal, &sa, NULL) != 0)
+        goto fail;
+
+    if (p->callback == NULL || p->callback == SIG_DFL)
+        how = SIG_UNBLOCK;
+    else if (p->callback == SIG_IGN)
+        how = SIG_BLOCK;
+    else
+        how = SIG_SETMASK;
+
+    if (p->callback == SIG_IGN) {
+        if (sigemptyset(&sigs) != 0 || sigaddset(&sigs, (int)p->signal) != 0)
+            goto fail;
+        if (sigprocmask(how, &sigs, NULL) != 0)
+            goto fail;
+    }
+
+    return 1;
+
+fail:
+    if (p->callback != NULL) {
+        p->callback = SIG_DFL;
+        CRYPTO_SIGNAL_block(p);
+    }
+    return 0;
+}
+
+#else
+
+static int CRYPTO_SIGNAL_arch_enable()
+{
+    return 0;
+}
+
+static int CRYPTO_SIGNAL_arch_disable()
+{
+    return 0;
+}
+
+static int CRYPTO_SIGNAL_block_arch(CRYPTO_SIGNAL *p)
+{
+    return 0;
+}
+
+#endif
+
+int CRYPTO_SIGNAL_block(CRYPTO_SIGNAL *p)
+{
+    struct sigentry *s = NULL;
+
+    if (siglist.prev == NULL || siglist.next == NULL)
+        list_init(&siglist);
+
+    if (list_empty(&siglist)) {
+        if (p->callback == NULL)
+            return 1;
+        else if (CRYPTO_SIGNAL_arch_enable() == 0)
+            return 0;
+    } else {
+        struct list *l;
+        if ((l = list_find(&siglist, siglist_sigeq, (void*)&p->signal)) != NULL)
+            s = container_of(l, struct sigentry, list);
```

Comment:
Is ```container_of``` portable to C89?

Reply:
moved the assignment out of the conditional, so that's resolved: https://github.com/ckalina/openssl/commit/2e7ee1dde85fc7c910f66a0a1d8b95c36ca1ea6d; will comment on `container_of` below